lib/utils/azure_queues.ts

```diff
@@ -0,0 +1,100 @@
+/**
+ * Implements utility functions for Azure Storage queues
+ */
+import * as winston from "winston";
+
+import { IContext } from "azure-functions-types";
+import { QueueService } from "azure-storage";
+
+// Array of intervals, in seconds, to delay message processing
+// The array *must* contain = (maxDequeueCount - 1) items
+// Any value must be less than 7 days (< 604800 seconds)
+// if (MAX_RETRIES < maxDequeueCount) - 1 there will be other extraneous
+//    #(maxDequeueCount - MAX_RETRIES) retries with the default visibilityTimeout
+//    before putting the message into the poison queue
+// if (MAX_RETRIES > maxDequeueCount - 1) the system will retry even after maxDequeueCount
+//    is reached and duplicate messages will be put into the poison queue
+export const RETRIES: ReadonlyArray<number> = [
```

Comment:
where do these numbers come from? truncated exp backoff is perfectly fine, see an example implementation [here](https://github.com/akka/akka/blob/master/akka-actor/src/main/scala/akka/pattern/BackoffSupervisor.scala#L149)

btw you should add a randomized factor to the delay - see page 285 of [this classic school text](http://www.uoitc.edu.iq/images/documents/informatics-institute/exam_materials/Computer%20Networks%20-%20A%20Tanenbaum%20-%205th%20edition.pdf) for an explanation

Reply:
Using Math.min(maxBackoff, Math.pow(10, retries)) was [my first try](https://github.com/gunzip/digital-citizenship-functions/blob/backoff/lib/created_message_queue_handler.ts#L104).

With this formula you get retries at [ 10, 100, 1000, 10000, 100000 ], so the last retry is at about 1.5 days from the moment the message has been enqueued. So I began tweaking the formula to get more retries, considering that we have a timespan of 7 days (max time_to_live) for each message. Finally I asked myself "how using a (tweaked) formula is better than directly specify the intervals directly ?" and went with the current implementation. 

The random delay is useful if you must avoid event collisions which is not our case IMHO: I'm not biased against adding a random factor, but note that tweaking the visibilityTimeout does not guarantee that the message will be scheduled at that exact time in the future, it only ensures that the message won't be processed *before the specified delay* (with some kind of imprecision dued to the Functions runtime). The moment the message will be dequeued depends on backpressure as well.

Reply:
![untitled](https://user-images.githubusercontent.com/48206/34778538-df838182-f61d-11e7-9b54-3f6b1ba41996.png)

Reply:
`300ms * 2^count` looks quite reasonable

Reply:
300ms * 2^count translates into 6 retries in the first 20 seconds (aka 6 more messages to process in the queue fo each message that end up with an error). IMHO this is way too much for the kind of errors that could cause a retry (network errors, paas service downtime, etc.) which are probably restored in seconds (if not minutes or hours). moreover this will pollute the queue with a lot of messages in case of subsequent errors. 

finally, another issue with any exponential strategy is that your last retry will be at best at 3.5 days from the time the message is accepted. this means, according to our max timespan of 7 days, you'll lose 3.5 days of possible retries (for example in case we opt for 300ms^2*count the last retry is at 3.6 days, then the message will expire)

Reply:
I think we are in the territory of "opinions" and unless you have some hard data to back your assumptions I would go for the [industry standard approach](https://docs.aws.amazon.com/general/latest/gr/api-retries.html) :)

Reply:
does `285ms * (2 ^ count)` look better?

![untitled](https://user-images.githubusercontent.com/48206/34780146-9e323ae8-f622-11e7-955c-fe51acc78f83.png)

Reply:
nicer ! BUT it looks like the visibilityTimeout granularity is in seconds (int):
http://azure.github.io/azure-storage-node/QueueService.html#updateMessage__anchor
so we must round the values or we have to chose a totally different way

Reply:
just round by excess using `Math.ceil()`

Reply:
I've plugged in the [backoff strategy](https://github.com/teamdigitale/digital-citizenship-functions/pull/133/files#diff-2d141546a695a0fa4638dedaf20abc72R23). 

Now we have ca. 30 retries. As maxDequeCount must be configured in host.json globally (once for every queue handler), it means that maxDequeCount must be 31, this in case you use a retry strategy OR NOT (so you'd better do : ). 

TL; DR: If you don't call retry() in your queue handlers they will retry enqueing the message 30 times in a very fast burst.